presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -598,17 +612,50 @@ else if (redistributeWrites) {
                     !source.getProperties().isCompatibleTablePartitioningWith(shufflePartitioningScheme.get().getPartitioning(), false, metadata, session) &&
                     !(source.getProperties().isRefinedPartitioningOver(shufflePartitioningScheme.get().getPartitioning(), false, metadata, session) &&
                             canPushdownPartialMerge(source.getNode(), partialMergePushdownStrategy))) {
+                PartitioningScheme exchangePartitioningScheme = shufflePartitioningScheme.get();
+                if (node.getTablePartitioningScheme().isPresent() && isPrestoSparkAssignBucketToPartitionForPartitionedTableWriteEnabled(session)) {
+                    int writerThreadsPerNode = getTaskPartitionedWriterCount(session);
```

Comment:
In Presto on Spark we always create 1 task per bucket. It is generally fine, as we can further partition data using local partitioning. However when writing to a bucketed table the data cannot be further partitioned, as all the data for a single bucket has to be written to a single file. Thus for the fragment that only contains a partitioned table writer operator we have to make sure that we assign as least as many buckets per task as there are threads available (`getTaskPartitionedWriterCount`).

What makes this PR special - is that the `bucketToPartition` is being assigned during the planning phase when usually in all other places it is assigned during the scheduling phase. This creates a precedent. Now the plan contains information about the physical partitions assignment, such as number of partitions and the mapping to buckets, before it get's to the scheduler (or RDD translator in case of Presto on Spark). 

Another approach would be to try to "reverse engineer" the plan during the scheduling phase (which in case of Presto on Spark is a translation to RDD). By "reverse engineer" I mean traversing the distributed plan and trying to deduce what fragment is a "partitioned table writer only", and then assign the `bucketToPartition` accordingly. 

I don't really like either of two solutions. I decided to go with the current approach as it is less likely to break. On the other side it breaks some assumptions that we currently have. 

I would really love to hear your thoughts.

CC: @highker @rongrong @shixuan-fan

Reply:
I don't think I fully understand the problem we are trying to solve so might ask some dumb questions. Is it fair to say that because we need to "assign as least as many buckets per task as there are threads available", we would need to assign `bucketToPartition` at planning phase so that we could do `addExchanges` accordingly?

Why do we need to assign at least as many buckets per task as there are threads available? My impression is that when creating writer, each writer would know the bucket they are trying to write, so ideally even if we have more threads than bucket, there should only be idle threads. What am I missing here?

Reply:
>  Is it fair to say that because we need to "assign as least as many buckets per task as there are threads available", we would need to assign bucketToPartition at planning phase so that we could do addExchanges accordingly?

Yeah, this is exactly what is happening. 

> Why do we need to assign at least as many buckets per task as there are threads available? My impression is that when creating writer, each writer would know the bucket they are trying to write, so ideally even if we have more threads than bucket, there should only be idle threads. What am I missing here?

It is usually not a problem in classic Presto, as it is multi-tenant. If one query doesn't use available CPU's - other queries will. Resource allocation in Spark works differently. In Spark we are being allocated a fixed number of CPU's. If we are not utilizing them - we are effectively wasting them.